src/lib/auth/auth.ts

```diff
@@ -0,0 +1,127 @@
+/* Facade for keycloak */
+import Keycloak, { KeycloakInitOptions, KeycloakInstance } from 'keycloak-js';
+
+const toPromise = keycloakPromise =>
+  new Promise<any>((resolve, reject) => keycloakPromise.success(resolve).error(reject));
+
+export interface ResourceRole {
+  resource: string;
+  resourceId: string;
+  role: string;
+}
+
+export interface AuthServiceConfiguration {
+  oidcIssuer: string;
+  clientId: string;
+  redirectUri: string;
+  logoutRedirectUri: string;
+}
+
+export interface User {
+  username: string;
+  name: string;
+}
+
+const extractUser: (string) => User = tokenParsed => ({ username: tokenParsed.user_name, name: tokenParsed.name });
+
+const extractResourceRoles: (authorities: string) => ResourceRole[] = (authorities: string) => {
+  return (authorities || '')
+    .split(',')
+    .map(authorityDescriptor => authorityDescriptor.split(':'))
+    .map(parts => ({ resource: parts[0], resourceId: parts[1], role: parts[2] }));
+};
+
+export class Auth {
+  private readonly kc: KeycloakInstance;
+
+  private readonly conf: AuthServiceConfiguration;
+
+  private initialized = false;
+
+  constructor(conf: AuthServiceConfiguration) {
+    const [url, realm] = conf.oidcIssuer.split('/realms/');
+
+    const kcConfig = {
+      realm,
+      url,
+      clientId: conf.clientId
+    };
+
+    this.conf = conf;
+    this.kc = Keycloak(kcConfig);
+  }
+
+  async init({ loginRequired }: { loginRequired: boolean }): Promise<boolean> {
+    const initOptions: KeycloakInitOptions = { onLoad: loginRequired ? 'login-required' : 'check-sso' };
+    await toPromise(this.kc.init(initOptions)).catch(console.error);
+    this.initialized = true;
+    return this.isAuthenticated();
+  }
+
+  onAuthenticationEvent(handler: (eventName: string) => void): void {
+    const bindEvents = [
+      'onReady',
+      'onAuthSuccess',
+      'onAuthError',
+      'onAuthRefreshSuccess',
+      'onAuthRefreshError',
+      'onAuthLogout',
+      'onTokenExpired'
+    ];
+    bindEvents.forEach(eventName => this.kc[eventName](() => handler(eventName)));
+  }
+
+  isInitialized: () => boolean = () => this.initialized;
+
+  login(): void {
+    this.kc.login();
+  }
+
+  logout(): void {
+    this.kc.logout({ redirectUri: this.conf.logoutRedirectUri });
+  }
+
+  isAuthenticated: () => boolean = () => this.kc.authenticated || false;
+
+  getUser(): User | null {
+    return (this.kc.tokenParsed && extractUser(this.kc.tokenParsed)) || null;
+  }
+
+  getToken: () => Promise<string | undefined> = async () => {
+    await toPromise(this.kc.updateToken(30)).catch(() => this.logout());
+    return this.kc.token;
+  };
+
+  getAuthorizationHeader: () => Promise<string> = async () => `Bearer ${await this.getToken()}`;
+
+  getAuthorities: () => string = () => this.kc.tokenParsed && (this.kc.tokenParsed as any).authorities;
```

Comment:
This function will not always return string. Undefined is also a possibility. What about a fallback to empty string?

Reply:
fikset, takk